<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8" />
    <title>rOpenSci | Rentrez 1_0 released</title>

     
    <script src="//cdn.bootcss.com/highlight.js/9.12.0/highlight.min.js"></script>
    <script src="//cdn.bootcss.com/highlight.js/9.12.0/languages/r.min.js"></script>

    <script type="text/javascript">
    hljs.configure({languages: []});
    hljs.initHighlightingOnLoad();
    </script>
    <link rel="stylesheet" type="text/css" href="/css/style-new.css" />
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="manifest" href="/manifest.json">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#4b8add">
<meta name="theme-color" content="#4b8add">


    <meta name="viewport" content="width=device-width, initial-scale=1">

<script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-462421-6', 'auto');
ga('send', 'pageview');
</script>

</head>
<body>

    
    <div id="header">
        <div class="container">
            <div class="row center-m">
                
                <div class="col-1">
                    <a href="/index.html"><img src="/img/icon_lettering_white.svg" /></a>
                </div>
                
                
                <nav class="col-6 col-offset-4 top-3">
                    <ul class="row between">
						<li><a href="/about/">About</a></li>
                        <li><a href="/blog/">Blog</a></li>
                        <li><a href="/packages/">Packages</a></li>
                        <li><a href="/community/">Community</a></li>
                        <li><a href="http://discuss.ropensci.org/" target="_blank">Discuss</a></li>

                    </ul>
                </nav>
                
            </div>
        </div>
    </div>


<article>
    <div class="container">
        <div class="row">
            <div class="col-2 sidebar top-5">
                <ul class="autoSidebar"></ul>
            </div> 
        </div>
        <div class="row">
            <div class="col-offset-2 col-8 top-20">
                <h2 style="text-align: center;">Rentrez 1_0 released</h2>
            </div>
        </div>
        <div class="row">
            <div class="col-offset-2 col-8 top-10 bottom-5" style="font-size: 20px; text-align: center;">
                <p class="authors" style="text-align: center;">
                    
                        
                          <a href="http://cartwrig.ht/people/#david-j-winter" target="_blank"> David Winter </a>&nbsp;
                        
                    
                </p>
                <br>
                <div style="font-size: 20px; color: #2C3E50; margin-top: 15px;">September 24, 2015</div>
            </div>
            <div class="col-offset-2 col-8 top-2">

                

                

<p>A new version of <code>rentrez</code>, our package for the NCBI&rsquo;s EUtils API, is making
it&rsquo;s way around the CRAN mirrors. This release represents a substantial
improvement to <code>rentrez</code>, including a <a href="https://cran.r-project.org/web/packages/rentrez/vignettes/rentrez_tutorial.html">new vignette</a>
that documents the whole package.</p>

<p>This posts describes some of the new things in <code>rentrez</code>, and gives us a chance
to thank some of the people that have contributed to this package&rsquo;s development.</p>

<h2 id="thanks">Thanks</h2>

<p>Thanks to everyone who has filed and issue or written us an email about <code>rentrez</code>,
your contributions have been an important part of the package&rsquo;s development. In particular, we welcome
<a href="http://ewrebio.me">Han Guangchun</a> as a new contributor to <code>rentrez</code> and thank
<a href="http://docking.org/~momeara/">Matthew O&rsquo;Meara</a> for posting
an issue that brought the <code>by_id</code> mode for <code>entrez_link</code>  (discussed below) to our
attention.</p>

<p>##The New Stuff</p>

<p>###Extract elements from the results of <code>entrez_summary()</code></p>

<p>The NCBI&rsquo;s &ldquo;summary records&rdquo; are very useful &ndash; they provide the most important
information about a given record in a relatively small and simple file. <code>rentrez</code>
provides the function <code>entrez_summary()</code> to retrieve these records. When more
than one unique ID is passed to <code>entrez_summary</code> the function returns a list of
esummary objects. For instance, you could find all the genetic variants associated
with asthma by finding links between the <a href="http://www.omim.org/entry/600807">OMIM record for asthma</a> and records in the database dbSNP:</p>

<pre><code class="language-r">snps &lt;- entrez_link(dbfrom=&quot;omim&quot;, db=&quot;snp&quot;, id= 600807)
snp_summs &lt;- entrez_summary(db=&quot;snp&quot;, id=snps$links$omim_snp)
</code></pre>

<p>A very common use-case for <code>entrez_summary()</code> is to extract a subset of the
elements from each record in that list. This release includes the function
<code>extract_from_esummary</code> to make this as straightforward as possible. It works with
a single element to extract:</p>

<pre><code class="language-r">extract_from_esummary(snp_summs, &quot;chr&quot;)
</code></pre>

<pre><code>## 11079657  2786098  1031772  1031771   545659
##     &quot;17&quot;      &quot;1&quot;      &quot;2&quot;      &quot;2&quot;     &quot;11&quot;
</code></pre>

<p>Or with multiple elements</p>

<pre><code class="language-r">summary_table &lt;- extract_from_esummary(snp_summs, c(&quot;chr&quot;, &quot;global_maf&quot;, &quot;fxn_class&quot;))
t(summary_table)
</code></pre>

<pre><code>##          chr  global_maf      fxn_class
## 11079657 &quot;17&quot; &quot;A=0.4295/2151&quot; &quot;intron-variant&quot;
## 2786098  &quot;1&quot;  &quot;T=0.1569/786&quot;  &quot;intron-variant&quot;
## 1031772  &quot;2&quot;  &quot;G=0.2131/1067&quot; &quot;downstream-variant-500B&quot;
## 1031771  &quot;2&quot;  &quot;T=0.2582/1293&quot; &quot;&quot;
## 545659   &quot;11&quot; &quot;C=0.3419/1712&quot; &quot;utr-variant-3-prime&quot;
</code></pre>

<p>###<code>entrez_link</code> can find external links</p>

<p>In addition to discovering links between records in NCBI databases, the function
<code>entrez_link</code> now provides support for finding external links (&lsquo;linkouts&rsquo; in
NCBI terminology). Perhaps the most interesting example is finding links for the
full text of articles in PubMed.</p>

<p>Let&rsquo;s try and find the full text of the paper describing <a href="https://github.com/ropensci/taxize">taxize</a> (using that article&rsquo;s PMID). To
override the functions default behaviour (finding links within NCBI databases)
we set the <code>cmd</code> argument to <code>llinks</code> (short for library links):</p>

<pre><code class="language-r">taxize_links &lt;- entrez_link(dbfrom=&quot;pubmed&quot;, id= 24555091, cmd=&quot;llinks&quot;)
taxize_links
</code></pre>

<pre><code>## elink object with contents:
##  $linkouts: links to external websites
</code></pre>

<p>The print function for this object tells you were the links live.</p>

<pre><code class="language-r">taxize_links$linkouts
</code></pre>

<pre><code>## $ID_24555091
## $ID_24555091[[1]]
## Linkout from F1000 Research Ltd
##  $Url: http://f1000research.com/a ...
##
## $ID_24555091[[2]]
## Linkout from Europe PubMed Central
##  $Url: http://europepmc.org/abstr ...
##
## $ID_24555091[[3]]
## Linkout from PubMed Central
##  $Url: http://www.ncbi.nlm.nih.go ...
##
## $ID_24555091[[4]]
## Linkout from PubMed Central Canada
##  $Url: http://pubmedcentralcanada ...
</code></pre>

<p>Each of those elements has a lot of information, but the URLs for each object
are probably the most important. For this reason, <code>rentrez</code> provides a function
to get just the URLs:</p>

<pre><code class="language-r">linkout_urls(taxize_links)
</code></pre>

<pre><code>## $ID_24555091
## [1] &quot;http://f1000research.com/articles/10.12688/f1000research.2-191.v2/doi&quot;
## [2] &quot;http://europepmc.org/abstract/MED/24555091&quot;
## [3] &quot;http://www.ncbi.nlm.nih.gov/pmc/articles/pmid/24555091/&quot;
## [4] &quot;http://pubmedcentralcanada.ca/pmcc/articles/pmid/24555091&quot;
</code></pre>

<p>###Web History features are easier to use</p>

<p>The NCBI provides a &ldquo;Web History&rdquo; feature to let users store the results of their
searches on the NCBI&rsquo;s severs and refer to those results without having to
pass unique ID&rsquo;s back and forth between computers. These features have always been
supported by <code>rentrez</code> but this release makes them easier to use.</p>

<p>Specifically, when the new optional argument <code>use_history</code> is set to <code>TRUE</code>
functions will return a <code>web_history</code> object which can be used in the place of unique
IDs in calls to <code>entrez_fetch</code>, <code>entrez_summary</code> or <code>entrez_link</code>.</p>

<p>To demonstrate, let&rsquo;s search for PubMed articles about the ciliate genus
<em>Tetrahymena</em>:</p>

<pre><code class="language-r">Tet_papers &lt;- entrez_search(db=&quot;pubmed&quot;, term=&quot;Tetrahymena[ORGN]&quot;, use_history=TRUE)
Tet_papers
</code></pre>

<pre><code>## Entrez search result with 6599 hits (object contains 20 IDs and a web_history object)
##  Search term (as translated):  &quot;tetrahymena&quot;[MeSH Terms] OR &quot;tetrahymena&quot;[All Fie ...
</code></pre>

<p>Now that we have a web_history object, we can use that to retrieve XML representations
of the first 10 records:</p>

<pre><code class="language-r">recs &lt;- entrez_fetch(db=&quot;pubmed&quot;,
                     web_history=Tet_papers$web_history,
                     retmax=10, rettype=&quot;xml&quot;)
</code></pre>

<p>###It&rsquo;s easier to keep track of which records are linked to other records</p>

<p>By default, when <code>entrez_link</code> gets a vector of more than one unique ID, it
returns sets of linked-IDs that match <em>any</em> of the IDs in the original call.
That means the user loses track of the mapping between the original IDs and those
from the linked database.</p>

<p>As of this release, <code>rentrez</code> supports the NCBI&rsquo;s <code>by_id</code> mode, which solves this problem.
Setting the new argument <code>by_id</code> to <code>TRUE</code> returns a list, with each element of
that list containing links for only one ID. To demonstrate, let&rsquo;s find protein
sequences associated with specific genes in the NCBI <code>gene</code> database:</p>

<pre><code class="language-r">all_links  &lt;- entrez_link(db=&quot;protein&quot;,
                          dbfrom=&quot;gene&quot;,
                          id=c(93100, 223646),
                          by_id=TRUE)

all_links
</code></pre>

<pre><code>## List of 2 elink objects,each containing
##   $links: IDs for linked records from NCBI
##
</code></pre>

<p>As you can see, printing the returned object let&rsquo;s you know what each element
contains, and you can extract the specific links you are looking for easily:</p>

<pre><code class="language-r">lapply(all_links, function(x) x$links$gene_protein)
</code></pre>

<pre><code>## [[1]]
##  [1] &quot;768043930&quot; &quot;767953815&quot; &quot;558472750&quot; &quot;194394158&quot; &quot;166221824&quot;
##  [6] &quot;154936864&quot; &quot;119602646&quot; &quot;119602645&quot; &quot;119602644&quot; &quot;119602643&quot;
## [11] &quot;119602642&quot; &quot;37787309&quot;  &quot;37787307&quot;  &quot;37787305&quot;  &quot;33991172&quot;
## [16] &quot;21619615&quot;  &quot;10834676&quot;
##
## [[2]]
##  [1] &quot;148697547&quot; &quot;148697546&quot; &quot;81899807&quot;  &quot;74215266&quot;  &quot;74186774&quot;
##  [6] &quot;37787317&quot;  &quot;37589273&quot;  &quot;31982089&quot;  &quot;26339824&quot;  &quot;26329351&quot;
</code></pre>

<h2 id="the-rest">The rest</h2>

<p>There are also numerous small changes that improve <code>rentrez</code>, fix bugs and
extend the package&rsquo;s documentation. We hope you find this new release helpful,
and as always we welcome bug reports via the <a href="https://github.com/ropensci/rentrez/issues">package&rsquo;s github repository</a>.</p>

            </div>
            <br>
            <div class="col-offset-2 col-8 top-4 labels">
                
                    <a href="/tags/r"><span class="label">R</span></a>
                
                    <a href="/tags/cran"><span class="label">CRAN</span></a>
                
                    <a href="/tags/rentrez"><span class="label">rentrez</span></a>
                
                    <a href="/tags/packages"><span class="label">packages</span></a>
                
                    <a href="/tags/ncbi"><span class="label">NCBI</span></a>
                
            </div>

            
            
            <div class="col-offset-2 col-8 top-4">
              <div id="disqus_thread"></div>
                <div id="disqus_thread"></div>
                <script type="text/javascript">
                var disqus_shortname = 'ropenscience';
                (function() {
                    var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
                    dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
                    (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
                })();
                </script>
                <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
                <a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
            </div>
            
        </div>
        <br>
    </div>
</article>





    <div id="footer">
        <div class="container">
            <div class="row start">
                <div class="col-1 col-1-3 top-8">
                    <img src="/img/icon_short_white.svg" />
                </div>

                <div class="col-9 col-offset-2 top-10 bottom-8">
                    <div class="row between">
                        <div class="col-2 col-1-3">
                            <a href="http://github.com/ropensci" target="_blank"><div class="icon icon-github"></div></a>
                            <a href="http://twitter.com/ropensci" target="_blank"><div class="icon icon-twitter"></div></a>
                            <a href="http://vimeo.com/ropensci" target="_blank"><div class="icon icon-vimeo"></div></a>
                        </div>
                    </div>
                    <div class="row">
                        <div class="col-2 col-1-3 top-8 bottom-8">
                            <ul>
                                <h5 class="bottom-2">Info</h5>
                                <li><a href="/about">Mission</a></li>
                                <li><a href="/about#team">Team</a></li>
                                <li><a href="/about#collaborators">Collaborators</a></li>
                                <li><a href="/careers">Careers</a></li>

                            </ul>
                        </div>

                        <div class="col-2 col-1-3 top-8 bottom-8">
                            <ul>
                                <h5 class="bottom-2">Work</h5>
                                <li><a href="/packages/">Packages</a></li>
                                <li><a href="/blog/">Blog</a></li>
                                <li><a href="/technotes/">Tech Notes</a></li>
                                <li><a href="/tutorials/">Tutorials</a></li>
                                <li><a href="/usecases/">Use Cases</a></li>
                                <li><a href="/resources/">More Resources</a></li>
                            </ul>
                        </div>
                        <div class="col-2 col-1-3 top-8 bottom-8">
                            <ul>
                                <h5 class="bottom-2">Participate</h5>
                                <li><a href="/contact.html">Contact us</a></li>
                                <li><a href="/community/">Community</a></li>
                                <li><a href="http://onboarding.ropensci.org/">Contribute software</a></li>
                                <li><a href="http://unconf17.ropensci.org/">Unconference</a></li>
                                <li><a href="/coc">Code of conduct</a></li>
                                <li><a href="/donate/">Donate</a></li>
                            </ul>
                        </div>
                        <div class="col-4 top-8 bottom-8">
                            <h5 class="bottom-2"></h5>

                            <p>rOpenSci is a fiscally sponsored project of <a href="http://numfocus.org" target="_blank">NumFOCUS</a></p>
                            <a href="http://numfocus.org" target="_blank"><img src="img/numfocus.png"></a>
                            <br>

                        </div>
                    </div>
                </div>
            </div>
            <div class="row">
                <div class="col-12 top-8 bottom-9 divider">
                    <p class="top-9">Except where otherwise noted, content on this site is licensed under the <a href="https://creativecommons.org/licenses/by/4.0/" target="_blank">CC-BY license</a>.
                    </p>
                </div>
            </div>
        </div>
    </div>



</body>
</html>



